### PR TITLE
CLN: Implement multiindex handling for get_op_result_name

### PR DESCRIPTION
**Pull request by phofl**
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

@jbrockmendel 

This would solve the inconsistency in ``get_op_result_name`` for MultiIndexes, but adds a bit more complexity here.
I think this would allow us to remove the now duplicated code in ``intersection``

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Can we do this by overriding get_reconciled_name_object and not get_op_result_name?  id like to avoid the small-but-frequent additional overhead there

**Comment by jreback** [comment]
pls merge master

**Comment by phofl** [comment]
Done and Done

File: pandas/core/ops/common.py [hunk 1]
```diff
@@ -122,3 +122,30 @@ def _maybe_match_name(a, b):
     elif b_has:
         return b.name
     return None
+
+
+def maybe_match_names_multiindex(a, b):
```
Comment (jreback) [inline pandas/core/ops/common.py:127]: can you type a, b and the return

File: pandas/core/ops/common.py [hunk 2]
```diff
@@ -93,7 +93,7 @@ def _maybe_match_name(a, b):
     """
```
Comment (jreback) [inline pandas/core/ops/common.py:93]: would rename this to maybe_match_names

File: pandas/core/ops/common.py [hunk 3]
```diff
@@ -122,3 +122,30 @@ def _maybe_match_name(a, b):
     elif b_has:
         return b.name
     return None
+
+
+def maybe_match_names_multiindex(a, b):
+    """
+    Try to find common names to attach to the result of an operation between
+    a and b.  Return a consensus list of names if they match at least partly
+    or None if they have completely different names.
```
Comment (jreback) [inline pandas/core/ops/common.py:131]: this could just be in padnas/core/indexes/multi.py i think would be fine

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
LGTM ex @jreback comments

**Comment by jreback** [comment]
also pls merge master

File: pandas/core/ops/common.py (hunk 1 above)
Comment (phofl) [reply pandas/core/ops/common.py]: Don't need to type this after moving?

File: pandas/core/ops/common.py (hunk 2 above)
Comment (phofl) [reply pandas/core/ops/common.py:93]: Done

File: pandas/core/ops/common.py (hunk 3 above)
Comment (phofl) [reply pandas/core/ops/common.py]: Done

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jreback** [comment]
looks good. is this user visible in any way? e.g you mention: `This would solve the inconsistency in get_op_result_name for MultiIndexes`

**Comment by phofl** [comment]
No, we did use rename previously which avoided this. But now we are consistent between the intersection methods

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**